File: backend/services/safety-gateway-platform/internal/integration/apollo_federation_client.go
package integration

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"go.uber.org/zap"
	"safety-gateway-platform/pkg/logger"
)

// ApolloFederationClient handles communication with Apollo Federation Gateway
type ApolloFederationClient struct {
	baseURL    string
	httpClient *http.Client
	logger     *logger.Logger
}

// GraphQLRequest represents a GraphQL request to Apollo Federation
type GraphQLRequest struct {
	Query     string                 `json:"query"`
	Variables map[string]interface{} `json:"variables,omitempty"`
}

// GraphQLResponse represents a GraphQL response from Apollo Federation
type GraphQLResponse struct {
	Data       map[string]interface{} `json:"data"`
	Errors     []GraphQLError         `json:"errors,omitempty"`
	Extensions map[string]interface{} `json:"extensions,omitempty"`
	DataSources []string              `json:"-"` // Populated from extensions
}

// GraphQLError represents a GraphQL error
type GraphQLError struct {
	Message    string                 `json:"message"`
	Locations  []GraphQLLocation      `json:"locations,omitempty"`
	Path       []interface{}          `json:"path,omitempty"`
	Extensions map[string]interface{} `json:"extensions,omitempty"`
}

// GraphQLLocation represents a location in a GraphQL query
type GraphQLLocation struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

// NewApolloFederationClient creates a new Apollo Federation client
func NewApolloFederationClient(baseURL string, logger *logger.Logger) (*ApolloFederationClient, error) {
	if baseURL == "" {
		return nil, fmt.Errorf("Apollo Federation URL is required")
	}

	httpClient := &http.Client{
		Timeout: 30 * time.Second,
		Transport: &http.Transport{
			MaxIdleConns:        100,
			MaxIdleConnsPerHost: 10,
			IdleConnTimeout:     90 * time.Second,
		},
	}

	return &ApolloFederationClient{
		baseURL:    baseURL,
		httpClient: httpClient,
		logger:     logger,
	}, nil
}

// Query executes a GraphQL query against Apollo Federation
func (c *ApolloFederationClient) Query(
	ctx context.Context,
	query string,
	variables map[string]interface{},
) (*GraphQLResponse, error) {
	c.logger.Debug("Executing Apollo Federation query",
		zap.String("base_url", c.baseURL),
	)

	// Create GraphQL request
	gqlRequest := &GraphQLRequest{
		Query:     query,
		Variables: variables,
	}

	// Marshal request to JSON
	requestBody, err := json.Marshal(gqlRequest)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal GraphQL request: %w", err)
	}

	// Create HTTP request
	req, err := http.NewRequestWithContext(
		ctx,
		http.MethodPost,
		c.baseURL+"/graphql",
		bytes.NewBuffer(requestBody),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create HTTP request: %w", err)
	}

	// Set headers
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "Safety-Gateway-Platform/2.0")

	// Add authentication if needed
	if authToken := c.getAuthToken(); authToken != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", authToken))
	}

	// Execute request
	startTime := time.Now()
	resp, err := c.httpClient.Do(req)
	queryDuration := time.Since(startTime)

	if err != nil {
		c.logger.Error("Apollo Federation query failed",
			zap.Error(err),
			zap.Duration("duration", queryDuration),
		)
		return nil, fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	// Check HTTP status
	if resp.StatusCode != http.StatusOK {
		c.logger.Error("Apollo Federation returned non-200 status",
			zap.Int("status_code", resp.StatusCode),
			zap.Duration("duration", queryDuration),
		)
		return nil, fmt.Errorf("Apollo Federation returned status %d", resp.StatusCode)
	}

	// Parse response
	var gqlResponse GraphQLResponse
	if err := json.NewDecoder(resp.Body).Decode(&gqlResponse); err != nil {
		return nil, fmt.Errorf("failed to decode GraphQL response: %w", err)
	}

	// Extract data sources from extensions
	if extensions := gqlResponse.Extensions; extensions != nil {
		if tracing, ok := extensions["tracing"].(map[string]interface{}); ok {
			if execution, ok := tracing["execution"].(map[string]interface{}); ok {
				if resolvers, ok := execution["resolvers"].([]interface{}); ok {
					dataSources := make(map[string]bool)
					for _, resolver := range resolvers {
						if r, ok := resolver.(map[string]interface{}); ok {
							if serviceName, ok := r["serviceName"].(string); ok {
								dataSources[serviceName] = true
							}
						}
					}
					
					// Convert to slice
					var sources []string
					for source := range dataSources {
						sources = append(sources, source)
					}
					gqlResponse.DataSources = sources
				}
			}
		}
	}

	// Check for GraphQL errors
	if len(gqlResponse.Errors) > 0 {
		c.logger.Warn("Apollo Federation returned GraphQL errors",
			zap.Int("error_count", len(gqlResponse.Errors)),
			zap.Duration("duration", queryDuration),
		)
		
		// Log each error
		for _, gqlError := range gqlResponse.Errors {
			c.logger.Warn("GraphQL error",
				zap.String("message", gqlError.Message),
				zap.Any("path", gqlError.Path),
			)
		}
	}

	c.logger.Debug("Apollo Federation query completed",
		zap.Duration("duration", queryDuration),
		zap.Int("data_sources", len(gqlResponse.DataSources)),
		zap.Bool("has_errors", len(gqlResponse.Errors) > 0),
	)

	return &gqlResponse, nil
}

// QueryPatientClinicalData executes a comprehensive patient data query
func (c *ApolloFederationClient) QueryPatientClinicalData(
	ctx context.Context,
	patientID string,
) (*GraphQLResponse, error) {
	query := `
		query GetPatientClinicalData($patientId: ID!) {
			patient(id: $patientId) {
				id
				
				# Demographics
				demographics {
					age
					ageInYears
					gender
					weight {
						value
						unit
					}
					height {
						value
						unit
					}
					bmi {
						value
						unit
					}
					ethnicity
					race
				}
				
				# Allergies and Intolerances
				allergies {
					id
					substance {
						code
						display
						system
					}
					severity
					reaction
					status
					verificationStatus
					recordedDate
					recorder {
						reference
						display
					}
				}
				
				# Active Conditions
				conditions {
					id
					code {
						coding {
							code
							display
							system
						}
						text
					}
					category
					severity {
						coding {
							code
							display
							system
						}
					}
					clinicalStatus
					verificationStatus
					onsetDateTime
					recordedDate
					asserter {
						reference
						display
					}
				}
				
				# Current Medications
				medications {
					id
					medicationCodeableConcept {
						coding {
							code
							display
							system
						}
						text
					}
					status
					effectiveDateTime
					effectivePeriod {
						start
						end
					}
					dosageInstruction {
						text
						timing {
							repeat {
								frequency
								period
								periodUnit
							}
						}
						doseAndRate {
							doseQuantity {
								value
								unit
								system
							}
						}
						route {
							coding {
								code
								display
								system
							}
						}
					}
					requester {
						reference
						display
					}
				}
				
				# Recent Lab Results
				labResults(recent: true, limit: 50) {
					id
					code {
						coding {
							code
							display
							system
						}
						text
					}
					status
					effectiveDateTime
					valueQuantity {
						value
						unit
						system
					}
					valueCodeableConcept {
						coding {
							code
							display
							system
						}
					}
					interpretation {
						coding {
							code
							display
							system
						}
					}
					referenceRange {
						low {
							value
							unit
						}
						high {
							value
							unit
						}
						text
					}
					performer {
						reference
						display
					}
				}
				
				# Vital Signs
				vitalSigns(recent: true, limit: 20) {
					id
					code {
						coding {
							code
							display
							system
						}
					}
					status
					effectiveDateTime
					valueQuantity {
						value
						unit
						system
					}
					component {
						code {
							coding {
								code
								display
								system
							}
						}
						valueQuantity {
							value
							unit
							system
						}
					}
				}
				
				# Recent Procedures
				procedures(recent: true, limit: 10) {
					id
					code {
						coding {
							code
							display
							system
						}
						text
					}
					status
					performedDateTime
					performedPeriod {
						start
						end
					}
					performer {
						actor {
							reference
							display
						}
					}
					outcome {
						coding {
							code
							display
							system
						}
					}
				}
				
				# Recent Encounters
				encounters(recent: true, limit: 5) {
					id
					class {
						code
						display
					}
					type {
						coding {
							code
							display
							system
						}
					}
					status
					period {
						start
						end
					}
					reasonCode {
						coding {
							code
							display
							system
						}
					}
					participant {
						individual {
							reference
							display
						}
						type {
							coding {
								code
								display
								system
							}
						}
					}
					serviceProvider {
						reference
						display
					}
				}
			}
			
			# Additional clinical context
			clinicalContext(patientId: $patientId) {
				riskFactors {
					category
					description
					severity
					evidence
				}
				clinicalSummary {
					activeProblems
					keyFindings
					riskAssessment
				}
			}
		}
	`

	return c.Query(ctx, query, map[string]interface{}{
		"patientId": patientID,
	})
}

// QueryKnowledgeBaseVersions retrieves current versions of all knowledge bases
func (c *ApolloFederationClient) QueryKnowledgeBaseVersions(ctx context.Context) (*GraphQLResponse, error) {
	query := `
		query GetKnowledgeBaseVersions {
			knowledgeBases {
				kb1_dosing {
					version
					lastUpdated
					description
				}
				kb3_guidelines {
					version
					lastUpdated
					description
				}
				kb4_safety {
					version
					lastUpdated
					description
				}
				kb5_ddi {
					version
					lastUpdated
					description
				}
				kb7_terminology {
					version
					lastUpdated
					description
				}
			}
		}
	`

	return c.Query(ctx, query, nil)
}

// QueryMedicationKnowledge retrieves medication-specific knowledge
func (c *ApolloFederationClient) QueryMedicationKnowledge(
	ctx context.Context,
	medicationCodes []string,
) (*GraphQLResponse, error) {
	query := `
		query GetMedicationKnowledge($medicationCodes: [String!]!) {
			medicationKnowledge {
				medications(codes: $medicationCodes) {
					code
					display
					drugClass
					mechanism
					indications
					contraindications
					warnings
					precautions
					dosageInformation {
						route
						form
						strength
						adultDosing {
							indication
							dose
							frequency
							duration
						}
						pediatricDosing {
							ageGroup
							dose
							frequency
							duration
						}
						renalAdjustment {
							creatinineClearance
							adjustment
						}
						hepaticAdjustment {
							severity
							adjustment
						}
					}
				}
				
				interactions(codes: $medicationCodes) {
					drug1 {
						code
						display
					}
					drug2 {
						code
						display
					}
					severity
					mechanism
					clinicalEffect
					management
					evidence {
						level
						source
						reference
					}
				}
				
				allergies(codes: $medicationCodes) {
					medication {
						code
						display
					}
					allergen
					crossReactivity
					severity
					manifestations
				}
			}
		}
	`

	return c.Query(ctx, query, map[string]interface{}{
		"medicationCodes": medicationCodes,
	})
}

// Health checks the health of the Apollo Federation Gateway
func (c *ApolloFederationClient) Health(ctx context.Context) error {
	query := `
		query Health {
			__schema {
				queryType {
					name
				}
			}
		}
	`

	response, err := c.Query(ctx, query, nil)
	if err != nil {
		return fmt.Errorf("health check failed: %w", err)
	}

	if len(response.Errors) > 0 {
		return fmt.Errorf("health check returned errors: %+v", response.Errors)
	}

	return nil
}

// getAuthToken retrieves authentication token (placeholder implementation)
func (c *ApolloFederationClient) getAuthToken() string {
	// In a real implementation, this would retrieve a JWT token
	// from environment variables, a secret store, or token cache
	return ""
}

// Close cleans up the client resources
func (c *ApolloFederationClient) Close() error {
	// Close any persistent connections
	c.httpClient.CloseIdleConnections()
	return nil
}

// GetBaseURL returns the base URL of the Apollo Federation Gateway
func (c *ApolloFederationClient) GetBaseURL() string {
	return c.baseURL
}

// SetTimeout sets the HTTP client timeout
func (c *ApolloFederationClient) SetTimeout(timeout time.Duration) {
	c.httpClient.Timeout = timeout
}

// EnableRetries enables retry logic for failed requests
func (c *ApolloFederationClient) EnableRetries(maxRetries int, backoffDuration time.Duration) {
	// This would implement retry logic with exponential backoff
	// For now, it's a placeholder
	c.logger.Info("Retry logic enabled",
		zap.Int("max_retries", maxRetries),
		zap.Duration("backoff_duration", backoffDuration),
	)
}

File: backend/services/context-service/app/clients/apollo_federation_client.py
"""
Apollo Federation GraphQL Client for Context Service.

This client allows the Context Service to query the Apollo Federation Gateway
to get data from microservices, following the correct architecture:

Context Service → Apollo Federation → Microservices
"""

import aiohttp
import json
import logging
from typing import Dict, Any, Optional, List
from datetime import datetime

logger = logging.getLogger(__name__)


class ApolloFederationClient:
    """
    GraphQL client for querying Apollo Federation Gateway from Context Service.
    
    This implements the correct architecture where Context Service acts as a consumer
    of Apollo Federation, not a provider to it.
    """
    
    def __init__(self, apollo_federation_url: str = "http://localhost:4000/graphql"):
        self.apollo_federation_url = apollo_federation_url
        self.session = None
        
    async def __aenter__(self):
        """Create HTTP session for GraphQL requests."""
        self.session = aiohttp.ClientSession()
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Close HTTP session."""
        if self.session:
            await self.session.close()
    
    async def get_patient_data(self, patient_id: str) -> Dict[str, Any]:
        """
        Get patient data from Patient Service via Apollo Federation.
        
        Args:
            patient_id: The patient ID
            
        Returns:
            Patient data from Patient Service
        """
        query = """
        query GetPatientData($patientId: ID!) {
            patient(id: $patientId) {
                id
                resourceType
                name {
                    given
                    family
                    use
                }
                gender
                birthDate
                address {
                    line
                    city
                    state
                    postalCode
                    country
                }
                telecom {
                    system
                    value
                    use
                }
            }
        }
        """
        
        variables = {"patientId": patient_id}
        
        try:
            logger.info(f"🔍 Querying Apollo Federation for patient data: {patient_id}")
            result = await self._execute_query(query, variables)
            
            if result and "patient" in result:
                logger.info(f"✅ Successfully retrieved patient data via Apollo Federation")
                return result["patient"]
            else:
                logger.warning(f"⚠️ No patient data found for ID: {patient_id}")
                return {}
                
        except Exception as e:
            logger.error(f"❌ Failed to get patient data via Apollo Federation: {e}")
            return {}
    
    async def get_patient_medications(self, patient_id: str) -> List[Dict[str, Any]]:
        """
        Get patient medications from Medication Service via Apollo Federation.

        Args:
            patient_id: The patient ID

        Returns:
            List of medications from Medication Service
        """
        query = """
        query GetPatientMedications($patientId: String!) {
            medications(patientId: $patientId) {
                id
                status
                medicationCodeableConcept {
                    coding {
                        system
                        code
                        display
                    }
                    text
                }
                dosageInstruction {
                    text
                }
                effectiveDateTime
            }
        }
        """
        
        variables = {"patientId": patient_id}
        
        try:
            logger.info(f"🔍 Querying Apollo Federation for patient medications: {patient_id}")
            result = await self._execute_query(query, variables)
            
            if result and "medications" in result:
                medications = result["medications"]
                logger.info(f"✅ Successfully retrieved {len(medications)} medications via Apollo Federation")
                return medications
            else:
                logger.warning(f"⚠️ No medications found for patient ID: {patient_id}")
                return []
                
        except Exception as e:
            logger.error(f"❌ Failed to get patient medications via Apollo Federation: {e}")
            return []
    
    async def get_patient_conditions(self, patient_id: str) -> List[Dict[str, Any]]:
        """
        Get patient conditions from Condition Service via Apollo Federation.
        
        Args:
            patient_id: The patient ID
            
        Returns:
            List of conditions from Condition Service
        """
        query = """
        query GetPatientConditions($patientId: String!) {
            conditionsByPatient(patientId: $patientId) {
                id
                clinicalStatus
                verificationStatus
                code {
                    coding {
                        system
                        code
                        display
                    }
                    text
                }
                onsetDateTime
                recordedDate
            }
        }
        """
        
        variables = {"patientId": patient_id}
        
        try:
            logger.info(f"🔍 Querying Apollo Federation for patient conditions: {patient_id}")
            result = await self._execute_query(query, variables)
            
            if result and "conditionsByPatient" in result:
                conditions = result["conditionsByPatient"]
                logger.info(f"✅ Successfully retrieved {len(conditions)} conditions via Apollo Federation")
                return conditions
            else:
                logger.warning(f"⚠️ No conditions found for patient ID: {patient_id}")
                return []
                
        except Exception as e:
            logger.error(f"❌ Failed to get patient conditions via Apollo Federation: {e}")
            return []
    
    async def get_patient_observations(self, patient_id: str) -> List[Dict[str, Any]]:
        """
        Get patient observations from Observation Service via Apollo Federation.
        
        Args:
            patient_id: The patient ID
            
        Returns:
            List of observations from Observation Service
        """
        query = """
        query GetPatientObservations($patientId: String!) {
            observationsByPatient(patientId: $patientId) {
                id
                status
                code {
                    coding {
                        system
                        code
                        display
                    }
                    text
                }
                valueQuantity {
                    value
                    unit
                    system
                }
                effectiveDateTime
                issued
            }
        }
        """
        
        variables = {"patientId": patient_id}
        
        try:
            logger.info(f"🔍 Querying Apollo Federation for patient observations: {patient_id}")
            result = await self._execute_query(query, variables)
            
            if result and "observationsByPatient" in result:
                observations = result["observationsByPatient"]
                logger.info(f"✅ Successfully retrieved {len(observations)} observations via Apollo Federation")
                return observations
            else:
                logger.warning(f"⚠️ No observations found for patient ID: {patient_id}")
                return []
                
        except Exception as e:
            logger.error(f"❌ Failed to get patient observations via Apollo Federation: {e}")
            return []
    
    async def _execute_query(self, query: str, variables: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """
        Execute a GraphQL query against Apollo Federation.
        
        Args:
            query: The GraphQL query string
            variables: Query variables
            
        Returns:
            Query result data or None if failed
        """
        if not self.session:
            raise RuntimeError("HTTP session not initialized. Use 'async with' context manager.")
        
        payload = {
            "query": query,
            "variables": variables
        }
        
        try:
            logger.debug(f"🔗 Executing GraphQL query to Apollo Federation: {self.apollo_federation_url}")
            
            async with self.session.post(
                self.apollo_federation_url,
                json=payload,
                headers={"Content-Type": "application/json"},
                timeout=aiohttp.ClientTimeout(total=30)
            ) as response:
                
                if response.status != 200:
                    logger.error(f"❌ Apollo Federation returned status {response.status}")
                    return None
                
                result = await response.json()
                
                if "errors" in result:
                    logger.error(f"❌ GraphQL errors from Apollo Federation: {result['errors']}")
                    return None
                
                if "data" not in result:
                    logger.error(f"❌ No data in Apollo Federation response")
                    return None
                
                logger.debug(f"✅ Successfully executed GraphQL query via Apollo Federation")
                return result["data"]
                
        except aiohttp.ClientTimeout:
            logger.error(f"❌ Timeout querying Apollo Federation")
            return None
        except Exception as e:
            logger.error(f"❌ Error executing GraphQL query: {e}")
            return None


# Singleton instance for use across the Context Service
_apollo_client_instance = None

def get_apollo_federation_client() -> ApolloFederationClient:
    """Get or create singleton Apollo Federation client."""
    global _apollo_client_instance
    if _apollo_client_instance is None:
        _apollo_client_instance = ApolloFederationClient()
    return _apollo_client_instance
